### PR TITLE
feat(rust, python): parse year-month-day as Datetime in slow-path

### LINKED ISSUE #7371: year-month-day can sometimes be parsed as Datetime, sometimes not
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

Parsing year-month-day as `Datetime` is allowed if:
1. the format is inferred
2. the format is specified and the fastpath is used

It's not allowed if:
3. the format is specified and the fastpath cannot be used

### Reproducible example

```python
In [2]: pl.Series(['01-02-2000']).str.strptime(pl.Datetime)  # autoinferred: works
Out[2]:
shape: (1,)
Series: '' [datetime[μs]]
[
        2000-02-01 00:00:00
]

In [3]: pl.Series(['13/Dec/2000']).str.strptime(pl.Datetime, '%d/%b/%Y')  # fastpath: works
Out[3]:
shape: (1,)
Series: '' [datetime[μs]]
[
        2000-12-13 00:00:00
]

In [4]: pl.Series(['13/December/2000']).str.strptime(pl.Datetime, '%d/%B/%Y')  # slowpath: fails
---------------------------------------------------------------------------
ComputeError: strict conversion to dates failed, maybe set strict=False
```

### Expected behavior

I think these should either all parse, or all fail?

### Installed versions

<details>

```
---Version info---
Polars: 0.16.11
Index type: UInt32
Platform: Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Python: 3.8.16 (default, Dec  7 2022, 01:12:06)
[GCC 11.3.0]
---Optional dependencies---
pyarrow: 11.0.0
pandas: 2.0.0rc0
numpy: 1.24.2
fsspec: <not installed>
connectorx: <not installed>
xlsx2csv: <not installed>
deltalake: <not installed>
matplotlib: 3.6.2
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #7371

### CODE REVIEW

File: polars/polars-time/src/chunkedarray/utf8/mod.rs [hunk 1]
```diff
@@ -517,10 +517,24 @@ pub trait Utf8Methods: AsUtf8 {
...
                                     *cache_map.entry(s).or_insert_with(|| {
-                                        NaiveDateTime::parse_from_str(s, &fmt).ok().map(func)
+                                        NaiveDateTime::parse_from_str(s, &fmt)
+                                            .ok()
+                                            .map(func)
+                                            .or_else(|| {
+                                                NaiveDate::parse_from_str(s, &fmt).ok().map(|nd| {
+                                                    func(nd.and_hms_opt(0, 0, 0).unwrap())
+                                                })
+                                            })
```
Comment (MarcoGorelli) [inline polars/polars-time/src/chunkedarray/utf8/mod.rs:527]: I've made this the same as it is in

https://github.com/pola-rs/polars/blob/1fe6430600a1dd2b085c7201833059066a1092c1/polars/polars-time/src/chunkedarray/utf8/infer.rs#L184-L193

Comment (ritchie46) [reply polars/polars-time/src/chunkedarray/utf8/mod.rs:527]: Maybe we can use that function, or factor out the code into a function?

Comment (MarcoGorelli) [reply polars/polars-time/src/chunkedarray/utf8/mod.rs]: yes, that's better, thanks!

**Comment by ritchie46** [comment]
Thanks!

**Pull request merged**